android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java

```diff
@@ -0,0 +1,1060 @@
+                                + decFor.format(location.getLatitude()) + ", "
+                                + decFor.format(location.getLongitude()));
+                        connection.postLocationUpdate(location);
+                    } else {
+                        Log.e(TAG, "No connection for postLocationUpdate()");
+                    }
+                    return null;
+                });
+    }
+
+    // (Guava EventBus Interface)
+    // This class encapsulates what an app might implement to watch for edge events. Not every event
+    // needs to be implemented. If you just want FindCloudlet, just @Subscribe to FindCloudletEvent.
+    //
+    class EdgeEventsSubscriber {
+
+        // Subscribe to error handlers!
+        @Subscribe
+        public void onMessageEvent(EdgeEventsConnection.EdgeEventsError error) {
+            Log.d(TAG, "EdgeEvents error reported, reason: " + error);
+            if (error.toString().equals("eventTriggeredButCurrentCloudletIsBest")) {
+                return;
+            }
+            meHelperInterface.showError(error.toString());
+        }
+
+        // Subscribe to FindCloudletEvent updates to appInst. Reasons to do so include
+        // the AppInst Health and Latency spec for this application.
+        @Subscribe
+        public void onMessageEvent(FindCloudletEvent findCloudletEvent) {
+            Log.i(TAG, "Cloudlet update, reason: " + findCloudletEvent.trigger);
+
+            // Connect to new Cloudlet in the event here, preferably in a background task.
+            Log.i(TAG, "FindCloudletEvent. Cloudlet: " + findCloudletEvent.newCloudlet);
+
+            someText = findCloudletEvent.newCloudlet.getFqdn();
+            meHelperInterface.showMessage("FindCloudletEvent. FQDN: "+someText);
+
+            Log.i(TAG, "Received: Server pushed a new FindCloudletReply to switch to: " + findCloudletEvent);
+            handleFindCloudletServerPush(findCloudletEvent);
+        }
+
+        void handleEventLatencyResults(AppClient.ServerEdgeEvent event) {
+            meHelperInterface.showMessage("Received: " + event.getEventType());
+            LocOuterClass.Statistics s = event.getStatistics();
+            DecimalFormat decFor = new DecimalFormat("#.##");
+            String message = "Latency results: min/avg/max/stddev="
+                    + decFor.format(s.getMin()) + "/" + decFor.format(s.getAvg()) + "/"
+                    + decFor.format(s.getMax()) + "/" + decFor.format(s.getStdDev());
+            meHelperInterface.showMessage(message);
+        }
+
+        void handleFindCloudletServerPush(FindCloudletEvent event) {
+            meHelperInterface.showMessage("Received: FindCloudletEvent");
+
+            // In this demo case, use our existing interface to display the newly selected cloudlet on the map.
+            mClosestCloudlet = event.newCloudlet;
+            onFindCloudlet(mClosestCloudlet);
+        }

```

Comment:
hmm. @lgarner Does he need this? Shouldn't the server push FindCloudletEvent be sent to the onMessageEvent?

Reply:
@franklin-huang-mobiledgex, Though originally written for the "give me all events" mode, this is also called from onMessageEvent(), so yes, it's still needed.